### PR TITLE
[Feature] Privileged api - backend only

### PR DESCRIPTION
**Pull request by petertgiles**
> [!Warning] 
The core functionality of this PR requires a manual change to test breaking features.  Manually back out the changes of [disable breaking changes](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10030/commits/8972a29259957da7c73f199ed888241e7f345f4b) to properly test the API changes.

🤖 Related to (but does not close) #9866

## 👋 Introduction

This branch adds a second route to getting to the graphql API and enforces using it for any operation requiring privileged permissions.  

## 🕵️ Details

It works by using custom routing for lighthouse to accept two different routes.  A middleware adds a request variable when it detects that the protected route is used.  And finally, a custom user checker ensures that Laratrust only allows the permission when the request variable is detected.

The `MakesGraphQLRequests` trait is unable to properly work when custom routing is enabled for lighthouse, like this PR uses.  The vast majority of the file diffs in this PR are just adding a trait to set which route to use in every test.

### Maintain Existing Functionality

- [ ] Building and running this branch "as-is" maintains current functionality without breaking anything.

### Add New Functionality

1. Back out the changes of [disable breaking changes](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10030/commits/8972a29259957da7c73f199ed888241e7f345f4b) to add new, breaking, functionality.
 
> [!TIP] 
Graphiql will only run on the regular, unprotected route.  To test the protected route you will have to use a different client like Altair, Postman, Edit&Resend, etc.

For the purpose of this PR: 
- a limited permission is one available to the base, guest, or applicant roles
- a privileged permission is one not available to the base, guest, or applicant roles
- the unprotected route is /graphql
- the protected route is /admin/graphql

- [ ] Any operation requiring a limited permissions only will succeed on either route
- [ ] Any operation requiring a privileged permission will fail on the uprotected route
- [ ] Any operation requiring a privileged permission can succeed on the protected route
- [ ] PHPUnit tests pass

### CODE REVIEW

File: api/tests/Unit/ProtectedRequestUserCheckerTest.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        $privilegedPermission = 'create-any-classification';
+
+        assertTrue($checker->currentUserHasPermission($privilegedPermission));
+    }
+
+    public function testCanNotUsePrivilegedPermissionUnprotected()
+    {
+        $this->markTestSkipped('Enable to test the route protection logic.');
+
+        // simulate a protected request context
```
Comment (petertgiles) [inline api/tests/Unit/ProtectedRequestUserCheckerTest.php:76]:
```suggestion
        // simulate an unprotected request context
```

File: api/tests/Unit/ProtectedRequestUserCheckerTest.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        // a privileged permission can be used only in a protected request
+        $privilegedPermission = 'create-any-classification';
+
+        assertFalse($checker->currentUserHasPermission($privilegedPermission));
+    }
+
+    public function testCanNotUseUnknownPermissionUnprotected()
+    {
+        // simulate a protected request context
```
Comment (petertgiles) [inline api/tests/Unit/ProtectedRequestUserCheckerTest.php:88]:
```suggestion
        // simulate an unprotected request context
```

**Review by petertgiles — COMMENTED** [review]
Copypasta to fix

File: api/app/Http/Middleware/ProtectedRequest.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ */
+class ProtectedRequest
+{
+    /**
+     * @return mixed Any kind of response
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        $path = $request->path();
+        if (str_starts_with($path, 'admin')) {
```
Comment (tristan-orourke) [inline api/app/Http/Middleware/ProtectedRequest.php:20]: Maybe instead use `$request->is('admin/*')`. See https://laravel.com/docs/11.x/requests#inspecting-the-request-path. I'm just thinking about how our files like `adminPage.js` got flagged by the firewall 😅

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Very cool! It looks good to me, and well tested. Just one minor thing about the url check which makes me nervous.

File: api/app/Http/Middleware/ProtectedRequest.php (hunk 3 above)
Comment (petertgiles) [reply api/app/Http/Middleware/ProtectedRequest.php]: Oh, very good catch!  That's a nice little matcher function.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
I found two things!

1. **authorizedToView** scopes! We skip checking policies on queries which are protected aby `authorizedToView` and `authorizedToViewBasicInfo`. Those queries can now return privileged information on the limited graphql endpoint. The scopes also need to check for `isProtectedRequest` and handle it appropriately.
2. graphiql. I think it would be a good idea to copy the graphiql config file into our config folder. That would allow us attach it to the /admin/graphql endpoint if we want, for local devleopment.

**Review by tristan-orourke — APPROVED** [review]
Everything seems to work for me now! I must not have had the project built correctly when I thought I found issues with the scope.

**Pull request merged**